lightning/src/offers/invoice_request.rs

```diff
@@ -0,0 +1,111 @@
+// This file is Copyright its original authors, visible in version control
+// history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Data structures and encoding for `invoice_request` messages.
+
+use bitcoin::hash_types::BlockHash;
+use bitcoin::secp256k1::PublicKey;
+use bitcoin::secp256k1::schnorr::Signature;
+use core::convert::TryFrom;
+use core::str::FromStr;
+use ln::features::OfferFeatures;
+use offers::{Offer, PayerTlvStream};
+use offers::merkle::SignatureTlvStream;
+use offers::offer::{OfferTlvStream, ParsedOffer};
+use offers::parse::{Bech32Encode, ParseError, SemanticError};
+
+///
+pub struct InvoiceRequest {
+	bytes: Vec<u8>,
```

Comment:
When responding to an `invoice_request` with an `invoice`, fields from both the `offer` and the `invoice_request` must be reflected.

Reply:
We could avoid duplicating the field by always storing the bytes in `Offer` instead of using an empty `Vec` there. I wasn't too thrilled about having it empty depending on which is the outermost message. Likewise, for when parsing an `Invoice`.  And then implement `AsRef<[u8]>` on all messages to pull it from the `Offer`.